Computer vision example with FiftyOne and Ultralytics

htahir1 opened the pull request:
## Add Computer Vision Example with YOLOv8, FiftyOne, and Interactive Web UI

This PR adds a comprehensive computer vision example to the ZenML repository, showcasing object detection with YOLOv8, FiftyOne dataset management, and an interactive web UI. The example demonstrates a complete annotation workflow loop: export → train → predict → import → analyze → visualize.

Key features:
- Complete training pipeline with YOLOv8 model and FiftyOne dataset integration
- Real-time inference service with sub-second latency using ZenML's deployment system
- Interactive web UI for uploading images or using URLs for instant object detection
- Custom materializer for YOLO models to handle serialization and versioning
- Full dataset-model lineage tracking and visual performance analysis

The PR also updates documentation links in the user guide to point to this new example instead of the previous one in zenml-projects repository.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [x] **IMPORTANT**: I made sure that my changes are reflected properly in the following resources:
  - [x] [ZenML Docs](https://docs.zenml.io)
  - [x] Dashboard: Needs to be communicated to the frontend team.
  - [x] Templates: Might need adjustments (that are not reflected in the template tests) in case of non-breaking changes and deprecations.
  - [x] [Projects](https://github.com/zenml-io/zenml-projects): Depending on the version dependencies, different projects might get affected.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

strickvl commented on examples/computer_vision/launch_fiftyone.py:
btw these are all methods we have on our annotator stack component. if you
wanted to go the extra mile you could add a fiftyone integration as an annotator
stack component and then you don't have to add this extra script...

htahir1 replied:
Ill maybe put all of these in a seperate class for a refactor later

strickvl commented on examples/computer_vision/README.md:
````diff
@@ -0,0 +1,160 @@
+**Test batch inference locally**:
+
+```bash
+python run.py --predict --image https://ultralytics.com/images/bus.jpg
````
does this work if yuo have a local image and a pipeline deployed to cloud? I
guess not since it's working with image paths and not raw bytes. I guess this is
why people want the `UploadFile` functionality?

strickvl commented on examples/computer_vision/README.md:
```diff
@@ -0,0 +1,160 @@
+
+**Use the ZenML Deployment Playground**
+
+The ZenML dashboard includes a built-in playground for deployed pipelines, allowing you to test your service directly from the UI. Navigate to your deployment in the dashboard, fill in the image URL and confidence threshold, and see real-time detection results with visualizations.
```
do you actually see the image with its results in the dashboard, too?

htahir1 replied:
Not in the zenml dashboard

strickvl commented on examples/computer_vision/steps/evaluate.py:
```diff
@@ -0,0 +1,86 @@
+    data_yaml = dataset_path / "dataset.yaml"
+
+    if not data_yaml.exists():
+        logger.warning(f"Dataset config not found at {data_yaml}")
```
Somehow wonder if it's not better to error out completely vs giving these null
results? I know all the models love to try/except everything so nothing ever
fails, but it feels like you'd want it to error out here? (unless this is caught
somewhere else?)

strickvl commented on examples/computer_vision/pipelines/hooks.py:
```diff
@@ -0,0 +1,80 @@
+        raise
+
+
+def cleanup_model() -> None:
```
maybe add a note that this method is just there for illustrative / educational
purposes only?

strickvl reviewed (commented):
Looks broadly fine to me, tbh. Some smaller comments made inline.

htahir1 commented:
@strickvl ive addressed all your comments including abstracting things into an annotator. I dont want to make this an official annotator yet, I have not even modeled it as a zenml stack component. But it should serve as a good starting point eventually

strickvl commented on examples/computer_vision/README.md:
```diff
@@ -102,17 +115,28 @@ computer_vision/
 
+### **FiftyOne Annotator Class**
+
+This example includes a `FiftyOneAnnotator` class that encapsulates all FiftyOne functionality in a clean, ZenML-style interface. **Crucially, it fixes the COCO class mapping issue** that causes catastrophic training failure when FiftyOne exports datasets with sequential IDs instead of proper COCO class IDs.
```
Not sure the 'COCO class mapping issue' needs so much space.
